Userland/Libraries/LibJS/Parser.cpp

```diff
@@ -440,7 +440,19 @@ class ScopePusher {
 
     void set_uses_this()
     {
-        for (auto scope_ptr = this; scope_ptr; scope_ptr = scope_ptr->m_parent_scope) {
+        auto const* closest_function_scope = last_function_scope();
+        if (!closest_function_scope || !closest_function_scope->m_is_arrow_function)
+            return;
+
+        for (auto* scope_ptr = this; scope_ptr; scope_ptr = scope_ptr->m_parent_scope) {
+            if (scope_ptr->m_type == ScopeType::Function)
+                scope_ptr->m_uses_this = true;
+        }
```

Comment:
I suppose we could break early in this loop if we encounter `scope_ptr->m_uses_this == true`